types/prop-types/index.d.ts

```diff
@@ -1,33 +1,49 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.2
+// TypeScript Version: 2.8
 
-export type Validator<T> = (object: T, key: string, componentName: string, ...rest: any[]) => Error | null;
+import { ReactNode, ReactElement } from 'react';
 
-export interface Requireable<T> extends Validator<T> {
-    isRequired: Validator<T>;
+export type IsOptional<T> = undefined | null extends T ? true : undefined extends T ? true : null extends T ? true : false;
+
+export type RequiredKeys<V> = { [K in keyof V]: V[K] extends Validator<infer T> ? IsOptional<T> extends true ? never : K : never }[keyof V];
+export type OptionalKeys<V> = Exclude<keyof V, RequiredKeys<V>>;
+export type InferPropsInner<V> = { [K in keyof V]: InferType<V[K]>; };
+
+export type Validator<T> = (props: object, propName: string, componentName: string, location: string, propFullName: string, _NOMINAL_TYPE_HACK_DO_NOT_USE?: T) => Error | null;
+
+export interface Requireable<T> extends Validator<T | undefined | null> {
+    isRequired: Validator<NonNullable<T>>;
 }
 
-export type ValidationMap<T> = {[K in keyof T]?: Validator<T> };
+export type ValidationMap<T> = { [K in keyof T]-?: Validator<T[K]> };
+// Workaround for type inference, ValidationMap<{}> does not work correctly because of conditional types
+export type AnyValidationMap = { [K in keyof any]: Validator<any> | Requireable<any> };
+
+export type InferType<V> = V extends Validator<infer T> ? T : any;
+export type InferProps<V> =
+    & InferPropsInner<Pick<V, RequiredKeys<V>>>
+    & Partial<InferPropsInner<Pick<V, OptionalKeys<V>>>>;
 
 export const any: Requireable<any>;
-export const array: Requireable<any>;
-export const bool: Requireable<any>;
-export const func: Requireable<any>;
-export const number: Requireable<any>;
-export const object: Requireable<any>;
-export const string: Requireable<any>;
-export const node: Requireable<any>;
-export const element: Requireable<any>;
-export const symbol: Requireable<any>;
-export function instanceOf(expectedClass: {}): Requireable<any>;
-export function oneOf(types: any[]): Requireable<any>;
-export function oneOfType(types: Array<Validator<any>>): Requireable<any>;
-export function arrayOf(type: Validator<any>): Requireable<any>;
-export function objectOf(type: Validator<any>): Requireable<any>;
-export function shape(type: ValidationMap<any>): Requireable<any>;
+export const array: Requireable<any[]>;
+export const bool: Requireable<boolean>;
+export const func: Requireable<(...args: any[]) => any>;
+export const number: Requireable<number>;
+export const object: Requireable<object>;
+export const string: Requireable<string>;
+export const node: Requireable<ReactNode>;
+export const element: Requireable<ReactElement<any>>;
+export const symbol: Requireable<symbol>;
+export function instanceOf<T>(expectedClass: new (...args: any[]) => T): Requireable<T>;
+export function oneOf<T>(types: T[]): Requireable<T>;
```

Comment:
I think `T` needs to be constrained here so `PropTypes.oneOf(['a', 'b'])` produces literal types instead of `string`. Probably warrants a test in any case.

Reply:
I'm unsure how to do this if you can offer any assistance. `oneOf` takes an array of anything and the runtime will just do a strict equality check on them. 

In TS 3.0 I know we can infer tuple types from rest args, but the prop-types package does not take rest args for `oneOf`, so as far as I can tell any array literals passed in just gets widened by the compiler prior to being parsed as a parameter type.

Reply:
I'm using TS 3.1-dev in my VSCode, and seeing this with the generic type constraint (though I may be misinterpreting what you're saying):
![image](https://user-images.githubusercontent.com/12239873/42860570-d592956c-8a1c-11e8-9478-5892a7a819d3.png)

As opposed to:
![image](https://user-images.githubusercontent.com/12239873/42860589-ef51216c-8a1c-11e8-8f29-2096644efb2c.png)
(which wouldn't be helpful anyways since it's not actually capturing the literal types)

Reply:
Does `T extends string[]` instead of `any[]` get somewhere?

Reply:
Doesn't seem to be the case:
![image](https://user-images.githubusercontent.com/12239873/42888954-efca634a-8a6e-11e8-9359-9a21e7f95beb.png)